# REEL VE KOMPLEKS AĞLARLA YAPILAN DİZİ SINIFLANDRIMA CV SONUÇLARI İLE İSTATİSTİKSEL KARŞILAŞTIRMA:
## Kompleks ve Reel Kodlamanın Karşılaştırılması (Kodlama Ekseni)
###### CV sonuçlarının toplanması: 
Çalışmada eğitilen on altı modelin çapraz doğrulama (CV) çıktı dosyalarından kat-başına doğruluk (acc) değerleri okunmuş; her dosya fold sütununa göre sıralanarak acc dizileri hizalanmıştır.
######  Eşleştirmenin kurulması: 
Kat bölünmesi ortak tohumla (SPLIT_SEED = 42) sabitlendiğinden, her yapılandırmanın k. katı aynı 40 proteini içermektedir. Ana hipotez doğrultusunda, aynı dizi temsili için kompleks ve reel ağlar eşleştirilmiştir. Üç mimari (çok katmanlı algılayıcı, 1B evrişimli, 2B evrişimli) ve ilgili dizi temsilleri boyunca toplam 14 yapılandırma (7 kompleks + 7 reel) 7 eşleştirilmiş karşılaştırma oluşturmuştur. Her karşılaştırma için kat-başına fark dizisi Fark = acc_kompleks(k) − acc_reel(k) hesaplanmıştır.
######  Betimleyici özetleme: 
Her karşılaştırmada iki ağın ortalama ± standart sapma değerleri, ham ortalama doğruluk farkı (puan cinsinden) ve kompleks ağın kaç katta üstün geldiği (kat-kazanım sayısı) raporlanmıştır.
######  Birincil test — Nadeau–Bengio düzeltmeli t: 
Fark dizisi üzerinde, eğitim kümelerinin örtüşmesinden kaynaklanan varyans yanlılığını telafi eden düzeltme terimiyle (1/K + n_test/n_train = 0,211) eşleştirilmiş t istatistiği hesaplanmış; serbestlik derecesi K − 1 = 9 alınarak iki yönlü p değeri elde edilmiştir.
######  Doğrulayıcı test — Wilcoxon işaretli-sıra: 
Aynı fark dizisi üzerinde, normallik varsayımı gerektirmeyen ve aykırı katlara dayanıklı Wilcoxon testi kesin (exact) yöntemle uygulanmış; birincil testle yön tutarlılığı denetlenmiştir. Kesin yöntem, küçük örneklem ve çok sayıda bağ (eşit fark) içeren bu veride sonucun deterministik olmasını sağlar.
######  Etki büyüklüğü: 
Her karşılaştırma için eşleştirilmiş Cohen's d (d_z = ortalama(d) / s_d) hesaplanmış ve çok küçük / küçük / orta / büyük ölçütlerine göre yorumlanmıştır.
######  Çoklu karşılaştırma düzeltmesi: 
Yedi karşılaştırma tek bir test ailesi kabul edilmiş ve birincil (NB) testin ham p değerlerine Holm–Bonferroni ardışık düzeltmesi (m = 7) uygulanmıştır. Ham p değerleri de raporlanmakla birlikte, nihai anlamlılık kararı düzeltilmiş p değerleri üzerinden verilmiştir.

## 2B KD EYSA & 2B RD EYSA & DVM & RO (2B Kompleks Temsil Kodlama)
######  Kapsam: 
Aynı 2B kompleks dizi temsilleri girdi olarak kullanılarak dört model —2B kompleks evrişimli ağ, 2B reel evrişimli ağ, Destek Vektör Makinesi (DVM) ve Rastgele Orman (RO)— karşılaştırılmıştır. DVM ve RO klasik makine öğrenmesi yöntemleri olduğundan, bu analiz bir kodlama hipotezi sınaması değil; önerilen yöntemin klasik yaklaşımlar arasındaki konumunu ve kompleks temsilin bu yöntemlerle kullanılabilirliğini ortaya koyan bağlamsal bir değerlendirmedir.
######  Birincil (omnibus) test — Friedman: 
Ölçümlerin aynı katlar üzerinde alındığı eşleştirilmiş yapı gereği, dört modelin başarım dağılımlarının özdeş olup olmadığı parametrik olmayan Friedman testiyle (bağ düzeltmeli) sınanmış; etki büyüklüğü Kendall uyum katsayısı (W) ile raporlanmıştır.
######  Post-hoc test — Nemenyi: 
Friedman testi anlamlı bulunduğunda, farkın hangi model çiftlerinden kaynaklandığını belirlemek için Nemenyi testi uygulanmış ve ortalama sıra farkları kritik farkla (CD) karşılaştırılmıştır. Omnibus test anlamlı olmasaydı post-hoc yorumlanmayacaktı.
######  Odaklı planlı kontrast: 
Kodlama karşıtlığını izole eden 2B KD EYSA – 2B RD EYSA karşılaştırması, önceden planlanmış tek bir kontrast olarak Nadeau–Bengio düzeltmeli t, doğrulayıcı Wilcoxon ve Cohen's d_z ile ayrıca sınanmıştır. Tek bir dizi kodlamasının farklı makine öğrenmesinin de bulunduğu modellere uygulanan bir çerçeve olduğundan, dörtlü karşılaştırma reel-kompleks çerçevesinden ayrı değerlendirilmiştir.

-NOT:Bu çalışmanın gerekçesi metodolojiktir: çapraz doğrulama katlamaları eğitim verisini paylaştığından tam bağımsız değildir ve sıradan eşleştirilmiş t-testinin bağımsızlık varsayımını kısmen ihlal eder; bu sınırlılığı gidermek için, bağımlılığı hesaba katan Nadeau–Bengio düzeltilmiş varyans kestirimi teste dâhil edilmiştir.

## KARŞILAŞTIRMLAR:
#### 1 AŞAMA:
- Tüm çalışmaların 10 Fold CV Acc. sonuçlarının rapor dosyalarından çekilmesi ( Onaltı model)
#### 2. AŞAMA:
- KOMPLEKS VE REEL YÖNTEMLERİN KARŞILAŞTIRILMASI (Kompleks ve Reel Kodlama)
1) KDYSA & RRDYSA - DNA
2) KDYSA & RRDYSA - Kodon
3) KDYSA & RRDYSA - AA

4) 1B KD EYSA & 1B RD EYSA - DNA
5) 1B KD EYSA & 1B RD EYSA - Kodon
6) 1B KD EYSA & 1B RD EYSA - AA

7) 2B KD EYSA & 2B RD EYSA - AA (2B Kompleks Temsil Kodlama)

#### 3. AŞAMA:
- Önerilen 2B kompleks temsil yönteminin; Karmaşık Değerli 2B Evrişimli Sinir Ağları, Reel Değerli 2B Evrişimli Sinir Ağları, Destek Vektör Makineleri ve Rastgele Orman makine öğrenimi yöntemlerindeki  proteinleri sınıflandırma performansları karşılaşttırmalı değerlendirilmelketedir

1) 2B KD EYSA & 2B RD EYSA & DVM & RO -AA (2B Kompleks Temsil Kodlama) 


###  1. AŞAMA: Tüm çalışmaların 10 Fold CV Acc. sonuçlarının rapor dosyalarından çekilmesi ( Onaltı modelin çalışmasI)
- 16 modele ait tüm çalışma 10 Fold CV Acc. sonuçları rapor olarak kaydedilikleri dosya bağlantı yolu ile toplu olarak çekilmektedir.

In [6]:
"""
CV accuracy sonuclarını rapor dosyasından yazdırılması

"""
import os
import glob
import json
import pandas as pd

# ============================================================
# AYAR
# ============================================================
KLASOR = r"C:\Users\matya"

# Gruplar: (baslik, [(etiket, yol), ...])  — D
GRUPLAR = [
    ("KDYSA ile DNA, Kodon ve AA Sınıflandırma Sonuçları", [
        ("KDYSA — DNA",   r"ciktilar_wirtinger_mlp_DNA_L900_npz\cv_10_fold_npz.xlsx"),
        ("KDYSA — Kodon", r"ciktilar_wirtinger_mlp_kodon_L300_npz\cv_10_fold_npz.xlsx"),
        ("KDYSA — AA",    r"ciktilar_wirtinger_mlp_aa_L300_npz\cv_10_fold_npz.xlsx"),
    ]),
    ("RDYSA ile DNA, Kodon ve AA Sınıflandırma Sonuçları", [
        ("RDYSA — DNA",   r"reel_dna_cv_10_fold_npz.xlsx"),
        ("RDYSA — Kodon", r"reel_kodon_cv_10_fold_npz.xlsx"),
        ("RDYSA — AA",    r"reel_aa_cv_10_fold_npz.xlsx"),
    ]),
    ("1B KDEYSA ile DNA, Kodon ve AA Sınıflandırma Sonuçları", [
        ("1B KDEYSA — DNA",   r"cv_10_fold_1dconv_dna.xlsx"),
        ("1B KDEYSA — Kodon", r"cv_10_fold_1dconv_kodon.xlsx"),
        ("1B KDEYSA — AA",    r"cv_10_fold_1dconv_aa.xlsx"),
    ]),
    ("1B RDEYSA ile DNA, Kodon ve AA Sınıflandırma Sonuçları", [
        ("1B RDEYSA — DNA",   r"reel_1dcnn_dna_cv_10_fold.xlsx"),
        ("1B RDEYSA — Kodon", r"reel_1dcnn_kodon_cv_10_fold.xlsx"),
        ("1B RDEYSA — AA",    r"reel_1dcnn_aa_cv_10_fold.xlsx"),
    ]),
    ("Kompleks Tabanlı 2B Protein-AA Temsillerinin\n2B KDEYSA & 2B RDEYSA & DVM & RO ile\nSınıflandırılması", [
        ("2B KDEYSA-AA", r"2d_kompleks_sonuc.xlsx"),
        ("2B RDEYSA-AA", r"2d_reel_sonuc.xlsx"),
        ("2B DVM-AA",    r"svm_sonuc.xlsx"),
        ("2B RO-AA",     r"rf_sonuc.xlsx"),
    ]),
]

TEST_ACC_ELLE = {
    # 
}
ACC_SUTUN, FOLD_SUTUN, BEKLENEN_KAT = "acc", "fold", 10
CIZGI, ALT = "=" * 60, "-" * 60

# ============================================================
# YARDIMCILAR
# ============================================================
def yok(v):
    return v is None or (isinstance(v, float) and pd.isna(v))

def dosya_bul(yol):
    aday = yol if os.path.isabs(yol) else os.path.join(KLASOR, yol)
    if os.path.isfile(aday):
        return aday
    vurus = glob.glob(os.path.join(KLASOR, "**", os.path.basename(yol)), recursive=True)
    return vurus[0] if vurus else None

def oku(dosya, etiket):
    """CV acc metriklerini toplar."""
    xl = pd.ExcelFile(dosya)
    cv = None; fold_df = None
    # 1) per-fold acc: 'acc' sutunu olan ILK sayfa
    for s in xl.sheet_names:
        df = xl.parse(s)
        kl = {c.lower().strip(): c for c in df.columns}
        if ACC_SUTUN in kl:
            if FOLD_SUTUN in kl:
                df = df.sort_values(kl[FOLD_SUTUN]).reset_index(drop=True)
            cv = df[kl[ACC_SUTUN]].to_numpy(dtype=float); fold_df = df
            break

    # 2) 'ozet' sayfasi (yeni sema): test_acc, sinif-bazli, cv fallback
    oz = None
    for s in xl.sheet_names:
        if s.lower() == "ozet":
            oz = xl.parse(s).iloc[0].to_dict(); break

    test_acc = None; pc = None; cv_mean = None; cv_std = None
    if oz is not None:
        if not yok(oz.get("test_acc")):
            test_acc = float(oz["test_acc"])
        d = {}
        for sinif, on in [("GPCR", "gpcr"), ("Kinase", "kinase")]:
            p = oz.get(f"test_{on}_precision")
            if not yok(p):
                d[sinif] = {"p": float(p),
                            "r": float(oz.get(f"test_{on}_recall")),
                            "f": float(oz.get(f"test_{on}_f1"))}
        if d:
            pc = d
        if cv is None and not yok(oz.get("cv_acc_ort")):
            cv_mean = float(oz["cv_acc_ort"])
            cv_std = None if yok(oz.get("cv_acc_std")) else float(oz["cv_acc_std"])

    # 3) test_acc fallback: fold sayfasindaki test sutunu / deney_meta.json / elle
    if test_acc is None and fold_df is not None:
        for c in fold_df.columns:
            cl = c.lower()
            if "test" in cl and "acc" in cl:
                v = pd.to_numeric(fold_df[c], errors="coerce").dropna()
                if len(v):
                    test_acc = float(v.iloc[0]); break
    if test_acc is None:
        meta = os.path.join(os.path.dirname(dosya), "deney_meta.json")
        if os.path.isfile(meta):
            try:
                dj = json.load(open(meta, encoding="utf-8"))
                for k, v in dj.items():
                    if "test" in k.lower() and "acc" in k.lower() and isinstance(v, (int, float)):
                        test_acc = float(v); break
            except Exception:
                pass
    if test_acc is None and TEST_ACC_ELLE.get(etiket) is not None:
        test_acc = float(TEST_ACC_ELLE[etiket])

    if cv is not None:
        cv_mean = float(cv.mean())
        cv_std = float(cv.std(ddof=1)) if len(cv) > 1 else None

    return {"cv": cv, "cv_mean": cv_mean, "cv_std": cv_std, "test_acc": test_acc, "pc": pc}

def baslik_yaz(metin):
    print(CIZGI)
    for satir in str(metin).split("\n"):
        print(f"# {satir}")
    print(CIZGI)

# ============================================================
# CALISTIR
# ============================================================
print(f"Arama koku: {KLASOR}\n")
W = max(len(et) for _, ms in GRUPLAR for et, _ in ms)
sonuclar = {}

# ---- 1) CV ACCURACY (gruplu) ----
print(CIZGI); print("# CV ACCURACY SONUÇLARI"); print(CIZGI); print()
for baslik, modeller in GRUPLAR:
    baslik_yaz(baslik)
    for etiket, yol in modeller:
        dosya = dosya_bul(yol)
        if dosya is None:
            print(f"{etiket:<{W}}  [dosya yok]"); continue
        try:
            r = oku(dosya, etiket)
        except Exception as e:
            print(f"{etiket:<{W}}  [HATA: {e}]"); continue
        sonuclar[etiket] = r
        if r["cv_mean"] is None:
            print(f"{etiket:<{W}}  [CV acc yok]"); continue
        std = f" ± {r['cv_std']:.4f}" if r["cv_std"] is not None else ""
        line = f"{etiket:<{W}}  CV acc = {r['cv_mean']:.4f}{std}"
        if r["cv"] is not None:
            line += "   [kat: " + " ".join(f"{a:.3f}" for a in r["cv"]) + "]"
            if len(r["cv"]) != BEKLENEN_KAT:
                line += f"  (!{len(r['cv'])} kat)"
        print(line)
    print()

    



Arama koku: C:\Users\matya

# CV ACCURACY SONUÇLARI

# KDYSA ile DNA, Kodon ve AA Sınıflandırma Sonuçları
KDYSA — DNA        CV acc = 0.9150 ± 0.0474   [kat: 0.925 0.925 0.875 0.875 0.925 0.925 0.950 0.825 1.000 0.925]
KDYSA — Kodon      CV acc = 0.8350 ± 0.0357   [kat: 0.825 0.825 0.825 0.800 0.825 0.850 0.875 0.775 0.850 0.900]
KDYSA — AA         CV acc = 0.8800 ± 0.0550   [kat: 0.800 0.875 0.875 0.925 0.950 0.875 0.850 0.825 0.975 0.850]

# RDYSA ile DNA, Kodon ve AA Sınıflandırma Sonuçları
RDYSA — DNA        CV acc = 0.9100 ± 0.0568   [kat: 0.950 0.900 0.800 0.950 0.875 0.975 0.925 0.900 0.975 0.850]
RDYSA — Kodon      CV acc = 0.8600 ± 0.0516   [kat: 0.900 0.875 0.800 0.900 0.925 0.800 0.850 0.875 0.900 0.775]
RDYSA — AA         CV acc = 0.9150 ± 0.0444   [kat: 0.950 0.950 0.875 0.925 0.950 0.900 0.950 0.825 0.950 0.875]

# 1B KDEYSA ile DNA, Kodon ve AA Sınıflandırma Sonuçları
1B KDEYSA — DNA    CV acc = 0.9500 ± 0.0441   [kat: 0.975 0.900 0.875 0.975 0.975 1.000 0.950 0.900 1.00

## 2. AŞAMA: Kompleks ve Reel Model Sonuçları  Eşleştirilmiş 10-fold CV acc karşılşatıması:
- Her cift icin:
  - Nadeau-Bengio duzeltilmiş eşleştirilmiş t-testi  (BIRINCIL TEST)
  - Wilcoxon isaretli sıra testi                     (dogrulayici)
  - Paired Cohen's d (dz) + ham puan farkı + kat-kazanım sayısı
Aile duzeyinde:
  - Holm-Bonferroni (birincil = NB t p'leri uzerinden)

NOT: Fold sırası özdeş (SPLIT_SEED=42).

In [7]:
"""
Kodlama ekseni (kompleks vs reel) — eslestirilmis 10-fold CV acc karsilastirmasi.
Reel ve kompleks tabanli calismalara ait 7 karsilastirma.
Her cift icin:
  - Nadeau-Bengio duzeltilmis eslestirilmis t-testi  (BIRINCIL)
  - Wilcoxon isaretli sira testi                     (dogrulayici)
  - Paired Cohen's d (dz) + ham puan farki + kat-kazanim sayisi
Aile duzeyinde:
  - Holm-Bonferroni (birincil = NB t p'leri uzerinden, m=7)

NOT: Fold sirasi tum dosyalarda ozdes (SPLIT_SEED=42).
DUZELTME: Wilcoxon 'exact' yontemine sabitlendi -> kucuk n ve baglarda deterministik
ve 4'lu koddaki ayni cift ile birebir tutarli sonuc verir.
"""

import warnings
import numpy as np
from scipy import stats

# ── CV parametreleri
N_TOTAL, K = 400, 10
N_TEST  = N_TOTAL // K          # 40
N_TRAIN = N_TOTAL - N_TEST      # 360
ALPHA   = 0.05
CORR    = (1 / K) + (N_TEST / N_TRAIN)   # 0.2111

# ── 14 dizi
KDYSA_DNA      = [0.9250, 0.9250, 0.8750, 0.8750, 0.9250, 0.9250, 0.9500, 0.8250, 1.0000, 0.9250]
KDYSA_KODON    = [0.8250, 0.8250, 0.8250, 0.8000, 0.8250, 0.8500, 0.8750, 0.7750, 0.8500, 0.9000]
KDYSA_AA       = [0.8000, 0.8750, 0.8750, 0.9250, 0.9500, 0.8750, 0.8500, 0.8250, 0.9750, 0.8500]
K1B_DNA        = [0.9750, 0.9000, 0.8750, 0.9750, 0.9750, 1.0000, 0.9500, 0.9000, 1.0000, 0.9500]
K1B_KODON      = [1.000, 1.000, 0.800, 0.975, 0.975, 0.950, 0.875, 0.875, 0.925, 0.975]
K1B_AA         = [0.9750, 0.9250, 0.9000, 0.9750, 0.9000, 0.9500, 0.9500, 0.9250, 0.9750, 0.9000]

RDYSA_DNA      = [0.9500, 0.9000, 0.8000, 0.9500, 0.8750, 0.9750, 0.9250, 0.9000, 0.9750, 0.8500]
RDYSA_KODON    = [0.9000, 0.8750, 0.8000, 0.9000, 0.9250, 0.8000, 0.8500, 0.8750, 0.9000, 0.7750]
RDYSA_AA       = [0.9500, 0.9500, 0.8750, 0.9250, 0.9500, 0.9000, 0.9500, 0.8250, 0.9500, 0.8750]
R1B_DNA        = [0.9500, 0.8750, 0.6750, 0.8750, 0.9500, 0.6500, 0.8500, 0.8250, 0.9000, 0.7500]
R1B_KODON      = [0.8500, 0.9000, 0.7750, 0.9000, 0.7250, 0.8750, 0.8750, 0.7750, 0.9000, 0.7750]
R1B_AA         = [0.9250, 0.9500, 0.8000, 0.9000, 0.9500, 0.8750, 0.9250, 0.8250, 0.9500, 0.9500]
K2B_KDEYSA_AA  = [0.975, 1.000, 1.000, 0.975, 1.000, 1.000, 0.975, 0.950, 0.950, 0.925]
R2B_RDEYSA_AA  = [0.975, 1.000, 0.975, 0.950, 0.975, 0.975, 0.975, 0.900, 0.950, 0.950]

# ── 7 karsilastirma: (kompleks, reel)
KARSILASTIRMALAR = {
    "KDYSA & RDYSA — DNA"          : (KDYSA_DNA,     RDYSA_DNA),
    "KDYSA & RDYSA — Kodon"        : (KDYSA_KODON,   RDYSA_KODON),
    "KDYSA & RDYSA — AA"           : (KDYSA_AA,      RDYSA_AA),
    "1B-KDEYSA & 1B-RDEYSA — DNA"  : (K1B_DNA,       R1B_DNA),
    "1B-KDEYSA & 1B-RDEYSA — Kodon": (K1B_KODON,     R1B_KODON),
    "1B-KDEYSA & 1B-RDEYSA — AA"   : (K1B_AA,        R1B_AA),
    "2B-KDEYSA & 2B-RDEYSA — AA"   : (K2B_KDEYSA_AA, R2B_RDEYSA_AA),
}

# ============================================================

def d_yorum(d):
    a = abs(d)
    return "cok kucuk" if a < 0.2 else "kucuk" if a < 0.5 else "orta" if a < 0.8 else "buyuk"

def wilcoxon_guvenli(c, r):
    """Sifir olmayan fark yoksa tanimsiz; aksi halde exact Wilcoxon (deterministik)."""
    if np.all(c == r):
        return float("nan"), 1.0
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")  # bag/sifir uyarilari bastirilir
        return stats.wilcoxon(c, r, alternative="two-sided", method="exact")

def karsilastir(kompleks, reel):
    c = np.asarray(kompleks, float); r = np.asarray(reel, float)
    d = c - r; dm = d.mean(); dv = d.var(ddof=1); ds = d.std(ddof=1)

    if dv == 0:
        t_nb, p_nb = 0.0, 1.0
    else:
        t_nb = dm / np.sqrt(dv * CORR)
        p_nb = 2 * stats.t.sf(abs(t_nb), df=K - 1)

    w, p_w = wilcoxon_guvenli(c, r)

    return dict(
        dm=dm, ds=ds, t_nb=t_nb, p_nb=p_nb, w=w, p_w=p_w,
        cohen=(dm / ds if ds > 0 else 0.0),
        kazanim=int((d > 0).sum()), beraberlik=int((d == 0).sum()),
        c_mean=c.mean(), c_sd=c.std(ddof=1), r_mean=r.mean(), r_sd=r.std(ddof=1),
    )

def holm(pvals):
    """Holm-Bonferroni; giris sirasinda duzeltilmis p dizisi doner."""
    m = len(pvals); order = np.argsort(pvals, kind="stable"); adj = np.empty(m); prev = 0.0
    for rank, idx in enumerate(order):
        prev = max(prev, min((m - rank) * pvals[idx], 1.0))
        adj[idx] = prev
    return adj

# ============================================================
# CALISTIR
# ============================================================
adlar  = list(KARSILASTIRMALAR)
sonuc  = {ad: karsilastir(*KARSILASTIRMALAR[ad]) for ad in adlar}
p_ham  = np.array([sonuc[ad]["p_nb"] for ad in adlar])
p_holm = holm(p_ham)

sep = "=" * 84
for ad in adlar:
    s = sonuc[ad]
    print(sep); print(f"{ad}   (kompleks vs reel)"); print(sep)
    print(f"Kompleks : {s['c_mean']:.4f} ± {s['c_sd']:.4f}")
    print(f"Reel     : {s['r_mean']:.4f} ± {s['r_sd']:.4f}")
    ber = f", {s['beraberlik']} beraberlik" if s['beraberlik'] else ""
    print(f"Ort. fark: {s['dm']:+.4f}  ({s['dm']*100:+.2f} puan)  |  kompleks {s['kazanim']}/{K} katta ustun{ber}")
    print(f"NB t     : t={s['t_nb']:+.3f}  df={K-1}  p(ham)={s['p_nb']:.4f}")
    if np.isnan(s['w']):
        print("Wilcoxon : tanimsiz (tum farklar sifir)")
    else:
        print(f"Wilcoxon : W={s['w']:.1f}  p={s['p_w']:.4f}  (exact)")
    print(f"Cohen dz : {s['cohen']:+.3f}  ({d_yorum(s['cohen'])} etki)")

print("\n" + sep)
print(f"AILE DUZEYI KARAR — Holm-Bonferroni  (m={len(adlar)} test, birincil=NB t, alpha={ALPHA})")
print(sep)
print(f"{'Karsilastirma':<32}{'fark(puan)':>12}{'kat':>7}{'p(ham)':>10}{'p(Holm)':>10}   karar")
print("-" * 90)
for i, ad in enumerate(adlar):
    s = sonuc[ad]
    karar = "anlamli" if p_holm[i] < ALPHA else "yeterli kanit yok"
    print(f"{ad:<32}{s['dm']*100:>+11.2f}{s['kazanim']:>5}/{K}{s['p_nb']:>10.4f}{p_holm[i]:>10.4f}   {karar}")
print("\nNot: Wilcoxon dogrulayici; aile karari birincil (NB) p uzerinden verildi.")

KDYSA & RDYSA — DNA   (kompleks vs reel)
Kompleks : 0.9150 ± 0.0474
Reel     : 0.9100 ± 0.0568
Ort. fark: +0.0050  (+0.50 puan)  |  kompleks 6/10 katta ustun
NB t     : t=+0.189  df=9  p(ham)=0.8541
Wilcoxon : W=23.5  p=0.7695  (exact)
Cohen dz : +0.087  (cok kucuk etki)
KDYSA & RDYSA — Kodon   (kompleks vs reel)
Kompleks : 0.8350 ± 0.0357
Reel     : 0.8600 ± 0.0516
Ort. fark: -0.0250  (-2.50 puan)  |  kompleks 4/10 katta ustun
NB t     : t=-0.704  df=9  p(ham)=0.4992
Wilcoxon : W=16.0  p=0.2754  (exact)
Cohen dz : -0.323  (kucuk etki)
KDYSA & RDYSA — AA   (kompleks vs reel)
Kompleks : 0.8800 ± 0.0550
Reel     : 0.9150 ± 0.0444
Ort. fark: -0.0350  (-3.50 puan)  |  kompleks 1/10 katta ustun, 4 beraberlik
NB t     : t=-1.372  df=9  p(ham)=0.2033
Wilcoxon : W=2.0  p=0.0715  (exact)
Cohen dz : -0.630  (orta etki)
1B-KDEYSA & 1B-RDEYSA — DNA   (kompleks vs reel)
Kompleks : 0.9500 ± 0.0441
Reel     : 0.8300 ± 0.1059
Ort. fark: +0.1200  (+12.00 puan)  |  kompleks 10/10 katta ustun
NB t     : 

#### Açıklama:
- Kodlama ekseni, üç ağ mimarisi ve üç dizi temsili boyunca yedi eşleştirilmiş karşılaştırmayla sınanmıştır. Her karşılaştırmada birincil test Nadeau–Bengio düzeltmeli eşleştirilmiş t-testi, doğrulayıcı test Wilcoxon işaretli-sıra testi, etki ölçüsü ise Cohen's d_z'dir; aile-düzeyi Tip-I hata denetimi için birincil (NB) olasılık değerlerine Holm–Bonferroni düzeltmesi (m = 7) uygulanmıştır. Bulgular, kompleks kodlamanın üstünlüğünün mimariden bağımsız evrensel bir etki olmadığını, mimariye bağlı olarak değiştiğini ortaya koymaktadır.

#### KDYSA & RDYSA:
- Bu mimaride kompleks kodlama lehine tutarlı bir üstünlük gözlenmemiştir. DNA temsilinde iki yöntem pratikte eşdeğerdir (ortalama fark +0,50 puan; d_z = 0,09, çok küçük etki; NB-t p = 0,854). Kodon ve amino asit (AA) temsillerinde fark sayısal olarak reel kodlama yönünde belirmiş (sırasıyla −2,50 ve −3,50 puan), AA temsilinde orta düzeyde bir etki büyüklüğüne ulaşmıştır (d_z = −0,63); ancak bu farkların hiçbiri istatistiksel anlamlılığa erişmemiştir (NB-t p ≥ 0,203; düzeltilmiş p = 1,000). Dolayısıyla düz ÇKA mimarisinde, kompleks kodlamanın reel kodlamaya üstün olduğuna dair yeterli istatistiksel kanıt elde edilememiştir.

#### 1B KD EYSA & 1B RD EYSA:
- Kompleks kodlama lehine en güçlü ve tutarlı kanıt bu mimaride elde edilmiştir. DNA temsilinde kompleks ağ, on katın tamamında (10/10) reel ağı geçmiş, +12,00 puanlık ortalama farkla büyük bir etki büyüklüğü sergilemiştir (d_z = 1,16); bu üstünlük hem düzeltilmemiş NB-t testinde (p = 0,032) hem de Wilcoxon testinde (p = 0,002) anlamlıdır. Kodon temsilinde de benzer biçimde güçlü bir üstünlük gözlenmiş (+10,0 puan; 9/10 kat; d_z = 1,251 büyük etki; Wilcoxon p = 0,0075), AA temsilinde ise orta düzeyde ve tutarlı bir eğilim (+3,25 puan; 7/10 kat; d_z = 0,56 orta ettki) saptanmıştır. Bu bulgular, evrişimli mimaride kompleks kodlamanın —özellikle DNA ve kodon temsillerinde— belirgin bir başarım kazancı sağladığına işaret etmektedir.

#### 2B KD EYSA & 2B RD EYSA:
- Bu mimaride kompleks kodlama sayısal olarak öne çıkmış; 2B KD EYSA modeli 0,9725 ± 0,0275 ortalama doğrulukla çalışmanın en yüksek başarımını elde etmiştir. Reel karşılığına göre +1,00 puanlık küçük bir üstünlük ve küçük bir etki büyüklüğü gözlenmiş (d_z = 0,41; kompleks 5/10 katta üstün, 3 katta eşit), ancak bu fark istatistiksel anlamlılığa ulaşmamıştır (NB-t p = 0,391). Her iki 2B modelinin de çok yüksek bir doğruluk tavanına ulaşması (tavan etkisi), yöntemler arasındaki farkın daralmasına katkıda bulunmuş olabilir.

#### Genel 
- Yedi karşılaştırmanın hiçbiri Holm–Bonferroni düzeltmesi sonrasında anlamlılığını korumamış; en güçlü sinyal olan 1B-DNA karşılaştırması dahi düzeltilmiş p = 0,226 değerine gerilemiştir. Bu durum, kısmen yedi testin tek bir aile altında düzeltilmesinden, kısmen de her karşılaştırmadaki sınırlı örneklem gücünden (K = 10 kat) kaynaklanmaktadır; küçük örneklemde çoklu-karşılaştırma cezası, güçlü etkileri dahi biçimsel anlamlılık eşiğinin altına çekebilmektedir. Bu nedenle bulgular yalnızca düzeltilmiş olasılık değerleri üzerinden değil; etki büyüklükleri, kat-kazanım oranları ve tutarlılık örüntüsü birlikte değerlendirilerek yorumlanmıştır. Bu bütüncül değerlendirmeye göre kompleks kodlamanın başarım katkısı mimariye bağımlıdır: düz ÇKA mimarisinde belirgin değilken, tek boyutlu evrişimli mimaride büyük etki büyüklükleri ve yüksek kat-tutarlılığıyla (DNA'da 10/10, kodon'da 8/10 kat) güçlü biçimde ortaya çıkmakta, iki boyutlu mimaride ise yüksek doğruluk tavanı altında küçük bir olumlu eğilim olarak kalmaktadır. Sonuç olarak, kompleks kodlamanın en belirgin katkısı evrişimli ağ mimarileriyle birlikte gözlenmekte; ancak mevcut örneklem büyüklüğü, bu eğilimleri aile düzeyinde biçimsel istatistiksel anlamlılığa taşıyacak güçten yoksun kalmaktadır.

## 3. ADIM:  2B KD EYSA& 2B RD EYSA & DVM &RO Karşılaştırılması: (Kompleks Tabanlı 2B Protein-A.A Temsil Sınıflandırma)
- Önerilen yeni yaklaşımın; Karmaşık Değerli 2B Evrişimli Sinir Ağları, Reel Değerli 2B Evrişimli Sinir Ağları, Destek Vektör Makineleri ve Rastgele Orman makine öğrenimi yöntemlerindeki sınıflandırmada  proteinleri temsil performansları karşılaşttırmalı değerlendirilmelketedir.
- İzole 2B kompleks–reel karşılaştırması birincil/doğrulayıcı analiz olarak kodlama hipotezinin yanıtını verirken, dörtlü karşılaştırma ikincil/bağlamsal bir analiz olarak yöntem konumlandırması ve temsilin kullanılabilirliği anlamını taşır.


In [5]:

"""
Dört modelin (2B KDEYSA, 2B RDEYSA, 2B DVM, 2B RO) 10-kat eşleştirilmiş
CV accuracy sonuçlarının istatistiksel karşılaştırması.

Tasarım: her kat (fold) dört modelin de AYNI doğrulama örnekleriyle ölçüldüğü
eşleştirilmiş bir bloktur (SPLIT_SEED=42). Dolayısıyla "3+ ilişkili grup" durumu.

Test paketi:
  1) Friedman omnibus (non-parametrik, birincil)  -> "en az bir model farklı mı?"
  2) Kendall's W (omnibus etki büyüklüğü)
  3) Nemenyi post-hoc (YALNIZ Friedman anlamlıysa yorumla) + kritik fark (CD)
  4) [Opsiyonel] Ana hipotez için odaklı karşılaştırma: KDEYSA vs RDEYSA
     (Nadeau-Bengio düzeltmeli t + standart t + Wilcoxon + Cohen's d_z + K/B/K)

Bağımlılık: numpy, scipy (>=1.7 ; studentized_range için).
"""

import numpy as np
from scipy import stats

# ============================================================
# 1) VERİ  (dosyadan okumak istersen bu sözlüğü doldurmayı değiştir)
# ============================================================
model_isimleri = ["2B KDEYSA-AA", "2B RDEYSA-AA", "2B DVM-AA", "2B RO-AA"]

acc = {
    "2B KDEYSA-AA": [0.975, 1.000, 1.000, 0.975, 1.000, 1.000, 0.975, 0.950, 0.950, 0.925],
    "2B RDEYSA-AA": [0.975, 1.000, 0.975, 0.950, 0.975, 0.975, 0.975, 0.900, 0.950, 0.950],
    "2B DVM-AA":    [0.950, 0.975, 0.975, 0.950, 0.975, 0.950, 1.000, 0.925, 0.925, 0.900],
    "2B RO-AA":     [0.975, 0.925, 0.925, 1.000, 0.950, 0.950, 0.925, 0.950, 0.925, 0.900],
}

# CV tasarım parametreleri (NB düzeltmesi için)
N_TRAIN = 360   # her katta eğitim örneği (400 - 40)
N_TEST  = 40    # her katta doğrulama örneği (400 / 10)
ALPHA   = 0.05

# Matris: satır = kat (N), sütun = model (k)
data = np.array([acc[m] for m in model_isimleri], dtype=float).T
N, k = data.shape
assert all(len(v) == N for v in acc.values()), "Tüm modeller aynı kat sayısına sahip olmalı."


def cizgi(baslik=""):
    print("=" * 60)
    if baslik:
        print(baslik)
        print("=" * 60)


# ============================================================
# 2) BETİMLEYİCİ İSTATİSTİKLER
# ============================================================
cizgi("BETİMLEYİCİ İSTATİSTİKLER (kat-başına acc)")
for m in model_isimleri:
    a = np.array(acc[m])
    print(f"{m:14s}  ort={a.mean():.4f}  std={a.std(ddof=1):.4f}  "
          f"min={a.min():.3f}  maks={a.max():.3f}")

# Ortalama sıralar (en yüksek acc = sıra 1; bağlar için ortalama sıra)
ranks = np.array([stats.rankdata(-row, method="average") for row in data])
avg_ranks = ranks.mean(axis=0)
print("\nOrtalama sıralar (1 = en iyi):")
sirali = sorted(zip(model_isimleri, avg_ranks), key=lambda x: x[1])
for m, r in sirali:
    print(f"  {m:14s}  ort. sıra = {r:.3f}")

# ============================================================
# 3) FRIEDMAN OMNIBUS + KENDALL'S W
# ============================================================
cizgi("FRIEDMAN OMNIBUS TESTİ (birincil)")
fried_stat, fried_p = stats.friedmanchisquare(*[acc[m] for m in model_isimleri])
kendall_W = fried_stat / (N * (k - 1))    # 0-1 arası uyum katsayısı
print(f"chi-kare = {fried_stat:.4f}   sd = {k-1}   p = {fried_p:.4f}   "
      f"(bağ düzeltmesi uygulandı)")
print(f"Kendall's W (etki büyüklüğü) = {kendall_W:.3f}   "
      f"(~0.1 zayıf, ~0.3 orta, ~0.5 güçlü)")
anlamli = fried_p < ALPHA
print(f"Karar (alpha={ALPHA}): "
      + ("ANLAMLI -> en az bir model farklı, post-hoc yorumlanabilir."
         if anlamli else
         "ANLAMSIZ -> post-hoc YORUMLANMAZ ('fark yok' değil, 'kanıt yetersiz')."))

# ============================================================
# 4) NEMENYI POST-HOC + KRİTİK FARK (CD)
#    (studentized_range ile; scikit-posthocs gerekmez)
# ============================================================
cizgi("NEMENYI POST-HOC (tüm ikili karşılaştırmalar)")
se_rank = np.sqrt(k * (k + 1) / (6.0 * N))
q_alpha = stats.studentized_range.ppf(1 - ALPHA, k, np.inf) / np.sqrt(2)
CD = q_alpha * se_rank
print(f"Kritik fark (CD, alpha={ALPHA}) = {CD:.3f}  (ort. sıra farkı bu değeri "
      f"geçen ikili anlamlıdır)\n")

if not anlamli:
    print("[UYARI] Friedman anlamlı değil; aşağıdaki p-değerleri yalnız "
          "bilgi amaçlı, resmî sonuç olarak raporlanmamalıdır.\n")

# İkili p-değeri matrisi
P = np.ones((k, k))
for i in range(k):
    for j in range(k):
        if i != j:
            dif = abs(avg_ranks[i] - avg_ranks[j])
            q = dif / se_rank * np.sqrt(2)          
            P[i, j] = stats.studentized_range.sf(q, k, np.inf)

# Matrisi yazdır
kisalt = [m.replace("2B ", "").replace("-AA", "") for m in model_isimleri]
print("İkili düzeltilmiş p-değerleri:")
print("            " + "".join(f"{c:>10s}" for c in kisalt))
for i in range(k):
    satir = f"{kisalt[i]:>10s}  "
    for j in range(k):
        satir += f"{P[i, j]:>10.4f}" if i != j else f"{'-':>10s}"
    print(satir)

print("\nİkili özet (ort. sıra farkı vs CD):")
for i in range(k):
    for j in range(i + 1, k):
        dif = abs(avg_ranks[i] - avg_ranks[j])
        isim_i, isim_j = kisalt[i], kisalt[j]
        karar = "ANLAMLI" if dif > CD else "anlamsız"
        print(f"  {isim_i:>8s} vs {isim_j:<8s}  |Δsıra|={dif:.3f}  "
              f"p={P[i, j]:.4f}  -> {karar}")

# ============================================================
# 5) [OPSİYONEL] ANA HİPOTEZ: KOMPLEKS vs REEL (KDEYSA vs RDEYSA)
#    Önceden planlanmış tek kontrast -> doğrudan test edilebilir.
# ============================================================
cizgi("ODAKLI KARŞILAŞTIRMA: KDEYSA vs RDEYSA (ana hipotez)")

A = np.array(acc["2B KDEYSA-AA"])   # 
B = np.array(acc["2B RDEYSA-AA"])   
d = A - B                           

mean_d = d.mean()
s_d = d.std(ddof=1)

# Kazanç / beraberlik / kayıp
kazanc = int(np.sum(d > 0)); berabere = int(np.sum(d == 0)); kayip = int(np.sum(d < 0))

# Standart eşleştirilmiş t
t_std = mean_d / (s_d / np.sqrt(N))
p_std = 2 * stats.t.sf(abs(t_std), N - 1)

# Nadeau-Bengio düzeltmeli t
corr = 1.0 / N + N_TEST / N_TRAIN
se_nb = np.sqrt(corr * d.var(ddof=1))
t_nb = mean_d / se_nb
p_nb = 2 * stats.t.sf(abs(t_nb), N - 1)

# Wilcoxon (sıfır farklar atılır -> etkin n düşer; exact -> deterministik)
n_eff = int(np.sum(d != 0))
try:
    import warnings
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")  # bağ/sıfır uyarıları bastırılır
        w_stat, w_p = stats.wilcoxon(A, B, method="exact")
except ValueError:
    w_stat, w_p = np.nan, np.nan

# Etki büyüklüğü
d_z = mean_d / s_d

print(f"Ham ortalama acc farkı  = {mean_d:+.4f}  ({mean_d*100:+.2f} puan)")
print(f"Kazanç/Beraberlik/Kayıp = {kazanc}/{berabere}/{kayip}  (toplam {N} kat)")
print(f"Cohen's d_z             = {d_z:.3f}")
print("-" * 60)
print(f"Standart eşleşmiş t     : t={t_std:.3f}  sd={N-1}  p={p_std:.4f}")
print(f"Nadeau-Bengio düz. t    : t={t_nb:.3f}  sd={N-1}  p={p_nb:.4f}   "
      f"(düzeltme çarpanı={corr:.3f})")
print(f"Wilcoxon işaretli-sıra  : W={w_stat:.3f}  p={w_p:.4f}  (etkin n={n_eff})")
print("-" * 60)
print("Yorum: NB-t, CV katlarının eğitim örtüşmesini telafi eder ve birincildir;")
print("Wilcoxon yön-doğrulayıcıdır.")
print("'n=10 sınırlılığı dikkate alınmalı' .")
cizgi()

BETİMLEYİCİ İSTATİSTİKLER (kat-başına acc)
2B KDEYSA-AA    ort=0.9750  std=0.0264  min=0.925  maks=1.000
2B RDEYSA-AA    ort=0.9625  std=0.0270  min=0.900  maks=1.000
2B DVM-AA       ort=0.9525  std=0.0299  min=0.900  maks=1.000
2B RO-AA        ort=0.9425  std=0.0290  min=0.900  maks=1.000

Ortalama sıralar (1 = en iyi):
  2B KDEYSA-AA    ort. sıra = 1.600
  2B RDEYSA-AA    ort. sıra = 2.300
  2B DVM-AA       ort. sıra = 3.000
  2B RO-AA        ort. sıra = 3.100
FRIEDMAN OMNIBUS TESTİ (birincil)
chi-kare = 10.1860   sd = 3   p = 0.0170   (bağ düzeltmesi uygulandı)
Kendall's W (etki büyüklüğü) = 0.340   (~0.1 zayıf, ~0.3 orta, ~0.5 güçlü)
Karar (alpha=0.05): ANLAMLI -> en az bir model farklı, post-hoc yorumlanabilir.
NEMENYI POST-HOC (tüm ikili karşılaştırmalar)
Kritik fark (CD, alpha=0.05) = 1.483  (ort. sıra farkı bu değeri geçen ikili anlamlıdır)

İkili düzeltilmiş p-değerleri:
                KDEYSA    RDEYSA       DVM        RO
    KDEYSA           -    0.6189    0.0725    0.0463
 

- Sonuçlar genel olarak 2B KDEYSA-AA modelinin en yüksek ortalama doğruluk değerine (0.9750) ve en iyi ortalama sıra değerine (1.600) sahip olduğunu göstermektedir. Friedman testi anlamlı bulunmuştur (χ2=10.186, p=0.0170; Kendall’s W=0.340). Buna göre, dört modelin performanslarının tamamen aynı olduğuna ilişkin hipotez reddedilmiş ve modeller arasında en az bir anlamlı performans farklılığı bulunduğu belirlenmiştir. Kendall’s W değeri bu farklılığın orta düzeyde bir etkiye sahip olduğunu göstermektedir.
- Nemenyi post-hoc analizinde ise bu genel farklılığın yalnızca 2B KDEYSA-AA ile 2B RO-AA arasındaki karşılaştırmada istatistiksel olarak anlamlı olduğu görülmüştür (p=0.0463). 2B KDEYSA-AA ile 2B RDEYSA-AA (p=0.6189) ve 2B DVM-AA (p=0.0725) arasındaki farklar ise %5 anlamlılık düzeyine ulaşmamıştır.
- Ana hipotez olan 2B KDEYSA-AA–2B RDEYSA-AA karşılaştırmasında, KDEYSA-AA ortalama olarak 1.25 yüzde puan daha yüksek doğruluk elde etmiş; 10 katın 5'inde üstün, 4'ünde eşit ve yalnızca 1'inde daha düşük performans göstermiştir. Cohen's dz=0.588 değeri farkın orta düzeyde bir etkiye sahip olduğunu göstermektedir. Bununla birlikte, çapraz doğrulama katları arasındaki bağımlılığı dikkate alan ve birincil doğrulayıcı test olarak kullanılan Nadeau–Bengio düzeltilmiş t-testi anlamlı değildir (p=0.2324). Dolayısıyla KDEYSA-AA'nın RDEYSA-AA'ya karşı gözlenen performans avantajı yönsel ve pratik açıdan dikkat çekici olmakla birlikte, birincil istatistiksel test açısından bu üstünlüğü doğrulayacak yeterli kanıt bulunmamaktadır. Wilcoxon testi anlamlı sonuç vermiş olsa da (p=0.0405), etkin örneklem büyüklüğünün 6 olması nedeniyle bu sonuç destekleyici/yön-doğrulayıcı olarak değerlendirilmelidir.
- Genel olarak 2B KDEYSA-AA en yüksek ortalama performansa sahip modeldir ve 2B RO-AA karşısında istatistiksel olarak anlamlı bir üstünlük göstermektedir. Buna karşılık 2B RDEYSA-AA'ya karşı gözlenen üstünlük henüz birincil doğrulayıcı test tarafından istatistiksel olarak kesin biçimde desteklenmemektedir. Sonuçların yalnızca 10 çapraz doğrulama katına dayanması ve Wilcoxon analizinde etkin örneklem büyüklüğünün 6 olması nedeniyle bulgular temkinli yorumlanmalıdır.